### ISSUE TITLE
Add support for printing test case results individually (But continuing to run in parallel)?

### ISSUE THREAD

**Issue by TysonAndre** [paratestphp/paratest#250]
If you have a large test suite, then you will have to wait for many minutes for it to finish, even when running in parallel.
If there's a crash/unhandled error in paratest (e.g. php7 Error failing to load dataProvider for a test class), then you won't see the tests which failed before that error, only the unhandled error.

Desired output format: Something which resembles the concatenation of multiple phpunit runs, in the order of test completions.

Does anything like this exist?

```
ParaTest using PHPUnit 6.2.1 by Sebastian Bergmann and contributors.

Result of yTest:
...............................................................  63 / 213 ( 29%)
............................................................... 126 / 213 ( 59%)
............................................................... 189 / 213 ( 88%)
........................                                        213 / 213 (100%)
OK (213 tests, 605 assertions)
--------------------------------------------------------------------------------
Result of xTest:
EE..                                                            4 / 4 (100%)

Time: 1.3 seconds, Memory: 12.00MB

There were 2 errors:

1) xTest::test_method
Error: Call to a member function someMethod() on null

/path/x.php:33
/path/x.php:49
/path/xTest.php:167

2) xTest::test_method2
Error: Call to a member function someMethod() on null

/path/x.php:33
/path/x.php:49
/path/xTest.php:206
```

**Comment by julianseeger** [comment]
I'm not sure whats the exact problem you want to address.
> If you have a large test suite, then you will have to wait for many minutes for it to finish, even when running in parallel.

If you want fast error messages on error (fail fast), there is an `--stop-on-failure` option. That will stop and print on the first failure (which includes errors).

> If there's a crash/unhandled error in paratest (e.g. php7 Error failing to load dataProvider for a test class), then you won't see the tests which failed before that error, only the unhandled error.

So not if a test fails but if paratest itself crashes? That's a tough one. It actually shouldn't crash. But in cases like failing to load dataProvider it would be nice if paratest wouldn't crash but handle the test as an error and continue.

> Desired output format: Something which resembles the concatenation of multiple phpunit runs, in the order of test completions.

Your output example prints one phpunit-like result output per test class. It should be doable to detect when all tests for a class have run. But I don't get the point. Take the example to an extreme and imagine you have even bigger TestClasses ... then you would want to get an output every half-class? ;)
Dou you want to start fixing the problems while the tests are still running and need the output for that?
In that case an alternative output like testdox-text could be of more use.
```
 [E] BankAccount: Balance is initially zero
    Error: Call to a member function someMethod() on null
 [x] BankAccount: Balance cannot become negative
 [E] BankAccount: test3
    Error: Call to a member function someMethod() on null
 [x] BankAccount: test4
```
something like that doesn't exist IIRC but would that solve your problem?

**Comment by TysonAndre** [comment]
> Do you want to start fixing the problems while the tests are still running and need the output for that?

That's what I want to do. testdox-text output seems useful

**Comment by julianseeger** [comment]
Sounds like a plan.
* this ticket is for the testdox-text-like output that prints the tests line by line and the error information directly following the test name
* a new issue for handling test-loading exceptions as test errors

**Comment by ChrisEdwards** [comment]
I need this feature, but for a very different reason. I would like faster feedback on failures when running in Jenkins.

We are using paratest to run our end-to-end functional tests. These tests each take a very long time to run. When paratest runs, it emits test results on the same line, adding a ., F, E, S, etc. to the line until it is full, then goes to the next line. The problem is that Jenkins does not add any of the output until the line is complete. Consequently, it can take 10 minutes to get ANY feedback on whether a test failed, even if it failed right away.

If paratest could write a result per line, then we could get rapid feedback on failures and waste less time waiting. 

We could enable --stop-on-failure, but we would often like the tests to complete to see what other issues they find while we start working on the one we saw first.

Furthermore, waiting until all tests finish to actually see WHAT failed is really difficult.  It is sometimes faster for us to run plain old phpunit single-threaded if we know there will be a failure and use the --test-dox output. I prefer to use paratest though.

**Comment by stevedesmond-ca** [comment]
Same as @ChrisEdwards -- I have a custom `Printer` that handles this for `phpunit`, but it doesn't look like the `--printer` flag is respected when using `--passthru`

**Comment by TysonAndre** [comment]
Is the printer open source?

**Comment by stevedesmond-ca** [comment]
There are a couple printers built in to PHPUnit -- here's a minimal repro:

`composer.json`:
```json
{
    "require-dev": {
        "phpunit/phpunit": "^9.0",
        "brianium/paratest": "^4.0"
    }
}
```

`ReproTests.php`:
```php
<?php

use PHPUnit\Framework\TestCase;

class ReproTests extends TestCase
{
    public function testCaseNumber1()
    {
        $this->assertTrue(true);
    }

    public function testCaseNumber2()
    {
        $this->assertFalse(false);
    }
}
```

`PHPUnit` output:
```
$ vendor/bin/phpunit ReproTests.php --printer PHPUnit\\Util\\TestDox\\CliTestDoxPrinter
PHPUnit 9.0.1 by Sebastian Bergmann and contributors.

Repro Tests
 ✔ Case number 1
 ✔ Case number 2

Time: 51 ms, Memory: 4.00 MB

OK (2 tests, 2 assertions)
```

`paratest` output:
```
$ vendor/bin/paratest ReproTests.php --passthru="--printer PHPUnit\\Util\\TestDox\\CliTestDoxPrinter"

Running phpunit in 4 processes with X:\paratest-printer-repro\vendor\phpunit\phpunit\phpunit

..

Time: 218 ms, Memory: 4.00 MB

OK (2 tests, 2 assertions)
```

**Comment by SteveDesmond-ca** [comment]
If this isn't otherwise prioritized, I can try to take a look in the next 2 months before the clock ticks down again.

**Comment by Slamdunk** [comment]
TestDox support added in https://github.com/paratestphp/paratest/pull/679
